### PR TITLE
make guassian get rid of LAPACK

### PR DESCRIPTION
**Pull request by MikeLing**
Those  function names like 'cblas_dger' doesn't make sense to me.

### CODE REVIEW

File: scripts/check_format.sh [hunk 1]
```diff
@@ -53,4 +53,4 @@ if [[ $# -ne 2 ]]; then
 fi
 
 # Run the check
-check_shogun_style ${1:-} ${2:-}
+check_shogun_style ${1:-} ${2:-}
```
Comment (vigsterkr) [inline scripts/check_format.sh]: do we need this?

File: src/shogun/distributions/Gaussian.cpp [hunk 2]
```diff
@@ -166,24 +166,28 @@ float64_t CGaussian::update_params_em(float64_t* alpha_k, int32_t len)
...
-
-				cov_sum[0]+=temp*alpha_k[j];
-				break;
-		}
+#ifdef HAVE_LAPACK
+			    cblas_dger(
+			        CblasRowMajor, num_dim, num_dim, alpha_k[j], v.vector, 1,
+			        v.vector, 1, (double*)cov_sum.matrix, num_dim);
+#else
+			    linalg::cblas_dger<float64_t>(alpha_k[j], v, v, cov_sum);
```
Comment (vigsterkr) [inline src/shogun/distributions/Gaussian.cpp]: please use a different name, dont use `cblas_dger`

File: src/shogun/distributions/Gaussian.cpp [hunk 3]
```diff
@@ -233,14 +238,19 @@ float64_t CGaussian::compute_log_PDF(SGVector<float64_t> point)
...
 	if (m_cov_type==FULL)
 	{
-		float64_t* temp_holder=SG_MALLOC(float64_t, m_d.vlen);
+		SGVector<float64_t> temp_holder(m_d.vlen);
+		temp_holder.zero();
+#ifdef HAVE_LAPACK
 		cblas_dgemv(CblasRowMajor, CblasNoTrans, m_d.vlen, m_d.vlen,
 					1, m_u.matrix, m_d.vlen, difference, 1, 0, temp_holder, 1);
+#else
+		linalg::cblas_dgemv<float64_t>(
```
Comment (vigsterkr) [inline src/shogun/distributions/Gaussian.cpp]: same here.... use a better name for the function `cblas_dgemv`

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 4]
```diff
@@ -691,6 +711,50 @@ SGMatrix<T> matrix_prod(SGMatrix<T>& A, SGMatrix<T>& B,
...
+ *
+ * @param scaling factor for vector ax
+ * @param vector a
+ * @param Whether to transpose matrix a
+ * @param vector x
+ * @param scaling factor for vector y
+ * @param vector y
+ */
+template <typename T>
+void cblas_dgemv(
```
Comment (MikeLing) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: @vigsterkr how about `vvpm` which means **vector** produce **vector** then **plus** **matrix** . I know it still looks weird, but I can't find a better name to represent this function does like **vector** produce **vector** then **plus** **matrix**

**Comment by MikeLing** [comment]
Ask review for this pr. I guess it's ready to reviewed except the function names :)

@vigsterkr Do you have time to take a look at this? Thank you!

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 5]
```diff
@@ -350,6 +350,26 @@ Container<T> add(Container<T>& a, Container<T>& b, T alpha=1, T beta=1)
 }
 
 /**
+ * Performs the operation A = alpha * x * y' + A
+ *
+ * @param scaling factor for vector x
```
Comment (vigsterkr) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: you  mean:
```
@param alpha scaling factor for vector x
```

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 6]
```diff
@@ -691,6 +711,50 @@ SGMatrix<T> matrix_prod(SGMatrix<T>& A, SGMatrix<T>& B,
 }
 
 /**
+ * Performs the operation y = \alpha ax + \beta y
+ * This function multiplies a * x (after transposing a, if needed)
+ * and multiplies the resulting matrix by alpha. It then multiplies vector y by
+ * beta. It stores the sum of these two products in vector y
+ *
+ * @param scaling factor for vector ax
```
Comment (vigsterkr) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: first word after the `@param` should be always the name of the variable in the function description

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 7]
```diff
@@ -691,6 +711,50 @@ SGMatrix<T> matrix_prod(SGMatrix<T>& A, SGMatrix<T>& B,
...
+ * @param vector a
```
Comment (vigsterkr) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: isn't this a matrix? (2nd parameter of the function)

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 8]
```diff
@@ -691,6 +711,50 @@ SGMatrix<T> matrix_prod(SGMatrix<T>& A, SGMatrix<T>& B,
...
+ * @param Whether to transpose matrix a
+ * @param vector x
```
Comment (vigsterkr) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: this should be
```
@param x vector
```

File: src/shogun/mathematics/linalg/LinalgNamespace.h [hunk 9]
```diff
@@ -691,6 +711,50 @@ SGMatrix<T> matrix_prod(SGMatrix<T>& A, SGMatrix<T>& B,
...
+	auto temp_vector = matrix_prod(a, x, transpose);
+	add(temp_vector, y, y, alpha, beta);
+}
+
+/**
+ * This function multiplies a * b and multiplies the resulting matrix by alpha.
+ * It then multiplies matrix c by beta. It stores the sum of these two products
+ * in matrix c.
+ *
+ * @param scaling factor for matrix a*b
```
Comment (vigsterkr) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: same here as above... the doxygen needs to be fixed

File: src/shogun/mathematics/linalg/LinalgNamespace.h (hunk 5 above)
Comment (MikeLing) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: ah, thank you.

File: src/shogun/mathematics/linalg/LinalgNamespace.h (hunk 4 above)
Comment (vigsterkr) [inline src/shogun/mathematics/linalg/LinalgNamespace.h]: ok so let's just have it without the `cblas_` prefix as @lisitsyn suggested... so `dgemv`

**Comment by vigsterkr** [comment]
@MikeLing cool let's see what the CI does.

**Pull request merged**